### PR TITLE
Amqp

### PR DESCRIPTION
**Pull request by paultheb**
Adding AMQP Producer, Consumer, and Unit Tests for both.

### CODE REVIEW

File: src/Consumer/Amqp.php [hunk 1]
```diff
@@ -1,21 +1,115 @@
...
+    protected $channel;
+    protected $queues = [];
+    protected $msgs = [];
+
+    public function __construct(\PhpAmqpLib\Channel\AMQPChannel $channel) {
+        $this->channel = $channel;
+    }
+
+    public function on($event, $name) {
+        if(!isset($this->queues[$exchange])) {
```
Comment (jdrich) [inline src/Consumer/Amqp.php]: Don't use isset here, because of the issue we had previously.

File: src/Consumer/Amqp.php [hunk 2]
```diff
@@ -1,21 +1,115 @@
...
+        $no_local = false,
+        $no_ack = false,
+        $exclusive = false,
+        $nowait = false,
+        $callback = null,
+        $ticket = null,
+        $arguments = []
+    ) {
+
+        $temp_msgs = $this->msgs;
```
Comment (jdrich) [inline src/Consumer/Amqp.php]: Why are you doing this?

File: src/Consumer/Amqp.php [hunk 3]
```diff
@@ -1,21 +1,115 @@
...
+        $default_callback = function($msg) use (&$temp_msgs) {
+            $temp_msgs[] = $msg;
+        };
+
+        $this->msgs = $temp_msgs;
+
+        if($callback != null) {
+            $user_callback = $callback;
+
+            $callback = function($msg) use ($default_callback) {
```
Comment (jdrich) [inline src/Consumer/Amqp.php]: You're going to have to pass $user_callback here. Did you write a test for this?

File: src/Consumer/Amqp.php (hunk 2 above)
Comment (paultheb) [inline src/Consumer/Amqp.php]: Cannot use $this as a lexical variable for closures.

**Pull request merged**